NativeAOT/win/arm64: Fix the reloc type typo

kunalspathak opened the pull request:
- There was a typo, where a relocation should be `IMAGE_REL_ARM64_SECREL_LOW12A` instead of `IMAGE_REL_ARM64_SECREL_LOW12L` . 
- The codegen was not handling the `add` operation correctly, so fixed that.
- Fixed the handling of the new relocs in superpmi.
- Added `Arm64` target in `reproNative.vcxproj` for developing/testing on arm64.

Fixes:  https://github.com/dotnet/runtime/issues/104500

kunalspathak committed:
Fix the reloc type typo

kunalspathak commented:
/azp run runtime-nativeaot-outerloop

VSadov commented:
The whole thing could be 

```asm
        adrp        $destReg, [LOW RELOC _tls_index]
        ldr         $TrashRegister32Bit, [$destReg, LOW RELOC _tls_index]

        ldr         $destReg, [xpr, #__tls_array]
        ldr         $destReg, [$destReg, $TrashRegister32Bit uxtw #3]

        add         $destReg, $destReg, #0, lsl #0xC
        RELOC       0xA, tls_InlinedThreadStatics                          ;; IMAGE_REL_ARM64_SECREL_HIGH12A
        add         $destReg, $destReg, #0, lsl #0
        RELOC       0x9, tls_InlinedThreadStatics                          ;; IMAGE_REL_ARM64_SECREL_LOW12A
```
Here we have a native pointer to the location where managed pointer to TLS root is stored. 
Nothing needs tracking yet.

```asm
        ldr         x19, [$destReg]
           ; gcrRegs +[x19]
```

Here we have a GC pointer to the managed TLS root

kunalspathak committed:
do not track tls_index reference

kunalspathak committed:
Update guid so the new collection contains correct reloc data

kunalspathak committed:
Merge remote-tracking branch 'origin/main' into aot-diffs

kunalspathak commented:
Fixed to not track the `tls_index`.

```
G_M37313_IG03:        ; bbWeight=1, gcrefRegs=0000 {}, byrefRegs=0000 {}, byref, isz
            ldr     x0, [xpr, #0x58]
            adrp    x1, [HIGH RELOC #0x4000000000420060]
            add     x1, x1, [LOW RELOC #0x4000000000420060]
            ldr     w1, [x1]
            ldr     x0, [x0, w1, UXTW #3]
            add     x1, x1, [LOW RELOC #0x4000000000420068]
            add     x1, x1, #0
            add     x0, x0, x1
            ldr     x19, [x0]
                             ; gcrRegs +[x19]
            cbz     x19, G_M37313_IG07
```

kunalspathak commented:
/azp run runtime-nativeaot-outerloop

VSadov commented:
for the
```asm
            add     x1, x1, [LOW RELOC #0x4000000000420068]
            add     x1, x1, #0
```

why the second add has no reloc? Is that just visualization?

kunalspathak committed:
fix the code for Add

kunalspathak committed:
remove old code and add comment

kunalspathak committed:
jit format

kunalspathak commented:
> for the
> 
> ```assembly
>             add     x1, x1, [LOW RELOC #0x4000000000420068]
>             add     x1, x1, #0
> ```
> 
> why the second add has no reloc? Is that just visualization?

yes, fixed it and also fixed the codegen. Now we generate this:

```asm
G_M37313_IG03:  ;; offset=0x001C
        F9402E40          ldr     x0, [xpr, #0x58]
        90000001          adrp    x1, [HIGH RELOC #0x4000000000420060]
        91000021          add     x1, x1, [LOW RELOC #0x4000000000420060]
        B9400021          ldr     w1, [x1]
        F8615800          ldr     x0, [x0, w1, UXTW #3]
        91400000          add     x0, x0, [HIGH RELOC #0x4000000000420068]
        91000000          add     x0, x0, [LOW RELOC #0]
        F9400013          ldr     x19, [x0]
        B40000F3          cbz     x19, G_M37313_IG07
```

kunalspathak commented:
![image](https://github.com/user-attachments/assets/3c853afa-cb95-46f4-a5d5-60f0014c9318)

kunalspathak commented:
/azp run runtime-nativeaot-outerloop

kunalspathak committed:
Add Arm64 target in reproNative.vcxproj

kunalspathak committed:
jit format

kunalspathak committed:
Merge remote-tracking branch 'origin/main' into aot-diffs

kunalspathak commented:
/azp run runtime-nativeaot-outerloop

jakobbotsch commented on src/coreclr/jit/helperexpansion.cpp:
```diff
@@ -595,7 +595,10 @@ bool Compiler::fgExpandThreadLocalAccessForCallNativeAOT(BasicBlock** pBlock, St
 
         CORINFO_CONST_LOOKUP tlsIndexObject = threadStaticInfo.tlsIndexObject;
 
+        // We want tls_index to get relocated and hence create IconHandle node,
+        // but it does not have GC ref, so reset the type to TYP_I_IMPL instead of TYP_REF
         GenTree* dllRef = gtNewIconHandleNode((size_t)tlsIndexObject.handle, GTF_ICON_OBJ_HDL);
+        dllRef->gtType  = TYP_I_IMPL;
```
What is this the value of? `GTF_ICON_OBJ_HDL` is for non-GC objects. The JIT can call various object related JIT-EE APIs on this value, so this should probably use a different handle type and not have this retyping. If you just want a generic handle type we have `GTF_ICON_CONST_PTR` for pointers to generic immutable data and `GTF_ICON_GLOBAL_PTR` for pointers to generic mutable data.

kunalspathak replied:
This is the value of `_tls_index`. If I keep it as `IconHandle()`, it's type is set to `TYP_REF` and we end up adding gc tracking for this, but since it is a non-gc native pointer, we don't want to do that. Also I wanted to make this as handle, so it can be relocatable. I will see if I can use a different handle. We set the type to `TYP_REF` only for `GTF_ICON_OBJ_HDL` and `TYP_I_IMPL` otherwise, so definitely a different handle type makes sense.

jakobbotsch commented on src/coreclr/jit/helperexpansion.cpp:
```diff
@@ -610,6 +613,14 @@ bool Compiler::fgExpandThreadLocalAccessForCallNativeAOT(BasicBlock** pBlock, St
         GenTree* tlsRootOffset = gtNewIconNode((size_t)tlsRootObject, TYP_INT);
         tlsRootOffset->gtFlags |= GTF_ICON_SECREL_OFFSET;
 
+        if (TargetArchitecture::IsArm64)
+        {
+            // for windows/arm64, the immediate constant should be contained because it gets
+            // generated as part of ADD instruction that consumes this constant. See
+            // emitIns_Add_Add_Tls_Reloc().
+            tlsRootOffset->SetContained();
+        }
```
This expansion runs way before lower, even before we have LIR... It' odd to be setting things as contained at this point. Can it be done in lowering instead? What would happen if some phase decided to break apart this constant and its parent between this point and lowering?

kunalspathak replied:
Yeah, I was not entirely sure the right place to set this, but probably should be able to do it in lowering.

kunalspathak committed:
review feedback

kunalspathak commented:
@VSadov @dotnet/jit-contrib PTAL

kunalspathak committed:
jit format

jakobbotsch commented on src/coreclr/jit/codegenarm64.cpp:
```diff
@@ -2764,6 +2758,18 @@ void CodeGen::genCodeForBinary(GenTreeOp* tree)
         genProduceReg(tree);
         return;
     }
+    else if (compiler->IsTargetAbi(CORINFO_NATIVEAOT_ABI) && TargetOS::IsWindows &&
+             op2->IsIconHandle(GTF_ICON_SECREL_OFFSET))
```
So this constant isn't considered a reloc?

kunalspathak replied:
It is and hence we add `EA_CNS_RELOC_FLG | EA_CNS_SEC_RELOC` flags below.

kunalspathak commented:
/azp run runtime-nativeaot-outerloop

VSadov commented:
failures in System.IO.Packaging are likely related to https://github.com/dotnet/runtime/pull/105160

VSadov commented:
otherwise windows-arm64 Release does not have any failures

VSadov commented:
I've looked at emitted code in few places and it looks correct.
A typical case:

```asm
S_P_CoreLib_System_Threading_Monitor__Enter:
  0000000140494090: A9BD7BFD  stp         fp,lr,[sp,#-0x30]!
  0000000140494094: A901D3F3  stp         x19,x20,[sp,#0x18]
  0000000140494098: F90017F5  str         x21,[sp,#0x28]
  000000014049409C: 910003FD  mov         fp,sp
  00000001404940A0: F9000BBF  str         xzr,[fp,#0x10]
  00000001404940A4: AA0003F3  mov         x19,x0
  // accessing TLS
  00000001404940A8: F9402E40  ldr         x0,[xpr,#0x58]
  00000001404940AC: F000C4C1  adrp        x1,uenum_close_ptr
  00000001404940B0: 91080021  add         x1,x1,#0x200
  00000001404940B4: B9400021  ldr         w1,[x1]
  00000001404940B8: F8615800  ldr         x0,[x0,w1 uxtw #3]
  00000001404940BC: 91400000  add         x0,x0,#0,lsl #0xC
  00000001404940C0: 91004000  add         x0,x0,#0x10
  00000001404940C4: F9400014  ldr         x20,[x0]
  00000001404940C8: B4000BB4  cbz         x20,000000014049423C
. . .
. . .
```

VSadov reviewed (commented):
LGTM. Thanks!!

EgorBo commented:
Sorry, I didn't realize your PR bumped the JIT-EE guid and merged a community PR with jit-ee guid bump. I think you can probably resolve the conflict and merge without waiting for CI..

kunalspathak committed:
Merge remote-tracking branch 'origin/main' into aot-diffs

kunalspathak commented:
/ba-g failures are unrelated
